packages/@aws-cdk/aws-ec2/lib/bastion-host.ts

```diff
@@ -0,0 +1,94 @@
+import { PolicyStatement } from "@aws-cdk/aws-iam";
+import { CfnOutput, Construct } from "@aws-cdk/core";
+import { AmazonLinuxGeneration, AmazonLinuxImage, InstanceClass, InstanceSize, InstanceType } from ".";
+import { Instance } from "./instance";
+import { ISecurityGroup } from "./security-group";
+import { IVpc, SubnetType } from "./vpc";
+
+/**
+ * Properties of the bastion host
+ *
+ * @experimental
+ */
+export interface BastionHostLinuxProps {
+
+  /**
+   * In which AZ to place the instance within the VPC
+   *
+   * @default - Random zone.
+   */
+  readonly availabilityZone?: string;
+
+  /**
+   * VPC to launch the instance in.
+   */
+  readonly vpc: IVpc;
+
+  /**
+   * The name of the instance
+   *
+   * @default 'BastionHost'
+   */
+  readonly instanceName?: string;
+
+  /**
+   * Use a public subnet instead of a private one.
+   * Set this to 'true' if you need to connect to this instance via the internet and cannot use SSM.
+   * You have to allow port 22 manually by using the connections field
+   *
+   * @default - false
+   */
+  readonly publicSubnets?: boolean;
+
+  /**
+   * Security Group to assign to this instance
+   *
+   * @default - create new security group with no inbound and all outbound traffic allowed
+   */
+  readonly securityGroup?: ISecurityGroup;
+
+  /**
+   * Type of instance to launch
+   * @default 't3.nano'
+   */
+  readonly instanceType?: InstanceType;
+
+}
+
+/**
+ * This creates a linux bastion host you can use to connect to other instances or services in your VPC.
+ * The recommended way to connect to the bastion host is by using AWS Systems Manager Session Manager.
+ *
+ * The operating system is Amazon Linux 2 with the latest SSM agent installed
+ *
+ * You can also configure this bastion host to allow connections via SSH
+ *
+ * @experimental
+ */
+export class BastionHostLinux extends Instance {
```

Comment:
Sorry for not seeing this before: I also agree with Elad that this host should probably *encapsulate* rather than *extend* `Instance`.

Reply:
Can you elaborate on why? I still think that a bastion host is a special instance and should derive from it, so users have all the options they have with an instance but with some defaults regarding the use case. I thought OOP is one of the benefits of CDK.

Reply:
We achieve OOP/polymorphism by implementing interfaces. Interfaces declare "a (this) can do the same thing as a (that)", regardless of the implementation. So if you want a Bastion Host to be-an Instance, then declare it like so:

```ts
class BastionHostLinux implements IInstance 
```

Inheriting from a class brings both the "(this) can do the same as (that)" relationship, but it also couples one implementation to the other, and can have unintended implications for the future extensibility of the classes involved (either a change in the implementation of `Instance` has an unintended side effect in the implementation of `BastionHost`, or we now face the limitation that `BastionHost` can only extend one class).

In some situations implementation sharing is a perfectly valid choice (which would usually be made to make it  *quicker* to implement a certain class since you don't have to type so much), but polymorphism through interface implementation is generally a safer choice in the long term, when it's hard to predict future changes.

In this case, I'd rather be safe than sorry later.

Reply:
Ok, I will change it and see what methods and fields I want to expose from the instance class.